<!DOCTYPE html>
<html>
<head>
	<meta charset='utf-8'>
	<title>GenomicFiles</title>
	<meta name="description" content="GenomicFiles">
	<meta name="author" content="Michael Love">
	<meta name="generator" content="slidify" />
	
	<!-- LOAD STYLE SHEETS -->
	<link rel="stylesheet" href="libraries/frameworks/html5slides/default/styles.css">
	<link rel="stylesheet" href="libraries/highlighters/highlight.js/css/tomorrow.css" />  <link rel="stylesheet" href = "assets/css/ribbons.css">
  
</head>
<body style='display: none'>
	<section class='slides layout-regular template-regular'>
     <article class="" id="slide-1" style="background:;">
  <h3></h3>
  <p><center></p>

<h1>GenomicFiles <a href="http://www.bioconductor.org/packages/release/bioc/html/GenomicFiles.html">v1.0</a></h1>

<p></center></p>

<p><center>
Valerie Obenchain, Michael Love, Martin Morgan
</center></p>

<p><center>
BioC2014: 1 August 2014
</center></p>

</article>

<article class="" id="slide-2" style="background:;">
  <h3>Example: epigenetic profiling of cell types</h3>
  <p><br></p>

<p><center>
<img src="wustl.png" alt="">
</center></p>

</article>

<article class="" id="slide-3" style="background:;">
  <h3>My solution in 2011</h3>
  <ul>
<li>BEDTools: reads in 200 bp windows for 100+ samples</li>
<li>Load bin-level counts into R as a large sparse matrix</li>
</ul>

<p><br></p>

<p>\[
\begin{array}{cccc}
. & . & 6 & 7 & . & 1 & . & . \\
. & 3 & . & . & . & . & . & . \\
. & . & . & . & . & . & 2 & . \\
. & . & . & . & . & . & . & . \\
3 & 9 & . & 5 & 2 & 8 & 1 & 6 \\
. & . & . & . & . & . & . & 1 \\
. & . & . & 5 & . & . & . & . \\
2 & . & . & . & . & . & . & . 
\end{array}
\]</p>

</article>

<article class="" id="slide-4" style="background:;">
  <h3>Limitations to this approach</h3>
  <p><br></p>

<ul>
<li>Matrix multiplication \(X^t X\) possible, but slow (2 hours)</li>
<li>Limits on functions: only sparse \(\rightarrow\) sparse</li>
<li>Must threshold low counts to stay &lt; 10<sup>9</sup></li>
<li>For mammalian genome, seemed limited to 100s of samples</li>
</ul>

<p><br></p>

<p><code>cannot allocate vector of length ...</code></p>

</article>

<article class="" id="slide-5" style="background:;">
  <h3>Views approach</h3>
  <p>Each range is said to represent a <em>view</em> onto the sequence.</p>

<pre><code class="r">library(IRanges)
(x &lt;- Rle(c(rep(0,5),1:5)))
</code></pre>

<pre><code>## numeric-Rle of length 10 with 6 runs
##   Lengths: 5 1 1 1 1 1
##   Values : 0 1 2 3 4 5
</code></pre>

<pre><code class="r">(v &lt;- Views(x, IRanges(4,7)))
</code></pre>

<pre><code>## Views on a 10-length Rle subject
## 
## views:
##     start end width
## [1]     4   7     4 [0 0 1 2]
</code></pre>

</article>

<article class="" id="slide-6" style="background:;">
  <h3>Operating on views</h3>
  <pre><code class="r">v[[1]] * 10
</code></pre>

<pre><code>## numeric-Rle of length 4 with 3 runs
##   Lengths:  2  1  1
##   Values :  0 10 20
</code></pre>

<pre><code class="r">viewApply(v, function(z) z * 10)
</code></pre>

<pre><code>## [[1]]
## numeric-Rle of length 4 with 3 runs
##   Lengths:  2  1  1
##   Values :  0 10 20
</code></pre>

</article>

<article class="" id="slide-7" style="background:;">
  <h3>Going back to epigenetic tracks</h3>
  <p>Want to load only one region at a time, no limit on # of files</p>

<p><br></p>

<p><center>
<img src="diagram.png" alt="">
</center></p>

</article>

<article class="" id="slide-8" style="background:;">
  <h3>Request to bioc-devel</h3>
  <p><br><br></p>

<p>Nov 2013 ML: &quot;BigWigViews would be incredibly useful for genomics applications
where we want to scan along the genome looking at lots of samples.&quot;</p>

<p>Dec 2013 VO: &quot;I&#39;ve started a package called GenomicFileViews ...
to provide infrastructure for parallel execution over a group of common file types&quot;</p>

</article>

<article class="" id="slide-9" style="background:;">
  <h3></h3>
  <pre><code class="r">library(GenomicFiles)
bwfv &lt;- BigWigFileViews(files, fileRange=ranges)
bwfv
</code></pre>

<pre><code>## BigWigFileViews dim: 6 ranges x 4 samples 
## names: test.bw test.bw.1 test.bw.2 test.bw.3 
## detail: use fileList(), fileSample(), fileRange(), ...
</code></pre>

</article>

<article class="" id="slide-10" style="background:;">
  <h3></h3>
  <pre><code class="r">se &lt;- coverage(bwfv)
class(se)
</code></pre>

<pre><code>## [1] &quot;SummarizedExperiment&quot;
## attr(,&quot;package&quot;)
## [1] &quot;GenomicRanges&quot;
</code></pre>

<pre><code class="r">assay(se)[4,1]
</code></pre>

<pre><code>## [[1]]
## RleList of length 1
## $chr2
## numeric-Rle of length 100 with 2 runs
##   Lengths:    50    50
##   Values : -0.75  -0.5
</code></pre>

</article>

<article class="" id="slide-11" style="background:;">
  <h3></h3>
  <pre><code class="r">se &lt;- summary(bwfv)
assay(se)
</code></pre>

<pre><code>##      [,1]   [,2]   [,3]   [,4]  
## [1,] -1     -1     -1     -1    
## [2,] -0.875 -0.875 -0.875 -0.875
## [3,] -0.75  -0.75  -0.75  -0.75 
## [4,] -0.625 -0.625 -0.625 -0.625
## [5,] 0.25   0.25   0.25   0.25  
## [6,] 0.375  0.375  0.375  0.375
</code></pre>

</article>

<article class="" id="slide-12" style="background:;">
  <h3>map / reduce</h3>
  <pre><code class="r">file &lt;- files[1]
range &lt;- ranges[4]
import(BigWigFile(file), which=range, as=&quot;Rle&quot;)[range]
</code></pre>

<pre><code>## RleList of length 1
## $chr2
## numeric-Rle of length 100 with 2 runs
##   Lengths:    50    50
##   Values : -0.75  -0.5
</code></pre>

<pre><code class="r">MAPPER = function(range, file, ...) {
  import(BigWigFile(file), which=range, as=&quot;Rle&quot;)[range][[1]]
}
</code></pre>

</article>

<article class="" id="slide-13" style="background:;">
  <h3>map / reduce</h3>
  <pre><code class="r">REDUCER = function(mapped, ...) {
  Reduce(&quot;+&quot;, mapped)
}
</code></pre>

</article>

<article class="" id="slide-14" style="background:;">
  <h3>map / reduce</h3>
  <pre><code class="r">library(BiocParallel)
register(SerialParam()) # MulticoreParam(workers=4)
res &lt;- reduceByRange(ranges, files, MAPPER, REDUCER)
length(res)
</code></pre>

<pre><code>## [1] 6
</code></pre>

<pre><code class="r">res[[4]]
</code></pre>

<pre><code>## numeric-Rle of length 100 with 2 runs
##   Lengths: 50 50
##   Values : -3 -2
</code></pre>

</article>

<article class="" id="slide-15" style="background:;">
  <h3>Files supported</h3>
  <ul>
<li>Bam</li>
<li>BigWig</li>
<li>FASTA</li>
<li>(tabix, VCF)</li>
</ul>

</article>

<article class="" id="slide-16" style="background:;">
  <h3>Next</h3>
  <ul>
<li>Profiling</li>
<li>Guide user wrt optimization of region size</li>
<li>Suggestions?</li>
</ul>

<p>Thanks:</p>

<ul>
<li>Valerie Obenchain</li>
<li>Martin Morgan</li>
<li>Rafael Irizarry</li>
<li>Kasper Hansen</li>
</ul>

</article>

  </section>
</body>
  <!-- LOAD JAVASCRIPTS  -->
	<script src='libraries/frameworks/html5slides/default/slides.js'></script>
	<!-- MathJax: Fall back to local if CDN offline but local image fonts are not supported (saves >100MB) -->
	<script type="text/x-mathjax-config">
	  MathJax.Hub.Config({
	    tex2jax: {
	      inlineMath: [['$','$'], ['\\(','\\)']],
	      processEscapes: true
	    }
	  });
	</script>
	<script type="text/javascript" src="http://cdn.mathjax.org/mathjax/2.0-latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
	<!-- <script src="https://c328740.ssl.cf1.rackcdn.com/mathjax/2.0-latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
	</script> -->
	<script>window.MathJax || document.write('<script type="text/x-mathjax-config">MathJax.Hub.Config({"HTML-CSS":{imageFont:null}});<\/script><script src="libraries/widgets/mathjax/MathJax.js?config=TeX-AMS-MML_HTMLorMML"><\/script>')
</script>
<!-- LOAD HIGHLIGHTER JS FILES -->
	<script src="libraries/highlighters/highlight.js/highlight.pack.js"></script>
	<script>hljs.initHighlightingOnLoad();</script>
	<!-- DONE LOADING HIGHLIGHTER JS FILES -->
	 
	</html>
